docs/proposals/templates.md

```diff
@@ -0,0 +1,576 @@
+List), plus a parameterization section.  Components reference the parameter by name and the value of the parameter is 
+substituted during a processing step, prior to submitting each component to the appropriate API endpoint for creation.
+
+The primary capability provided is that parameter values can easily be shared between components, such as a database password 
+that is provided by the user once, but then attached as an environment variable to both a database pod and a web frontend pod.
+
+In addition, the template can be repeatedly instantiated for a consistent application deployment experience in different 
+namespaces or Kubernetes clusters.
+
+Lastly, we propose the Template API object include a “Labels” section in which the template author can define a set of labels 
+to be applied to all objects created from the template.  This will give the template deployer an easy way to manage all the 
+components created from a given template.  These labels will also be applied to selectors defined by Objects within the template,
+allowing a combination of templates and labels to be used to scope resources within a namespace.  That is, a given template
+can be instantiated multiple times within the same namespace, as long as a different label value is used each for each
+instantiation.  The resulting objects will be independent from a replica/load-balancing perspective.
+
+Generation of parameter values for fields such as Secrets will be delegated to an [admission controller/initializer/finalizer](https://github.com/kubernetes/kubernetes/issues/3585) rather than being solved by the template processor.
+
+Labels to be assigned to all objects could also be generated in addition to, or instead of, allowing labels to be supplied in the 
+Template definition.
+
+### API Objects
+
+**Template Object**
+```
+// Template contains the inputs needed to produce a Config.
+type Template struct {
+    unversioned.TypeMeta
+    kapi.ObjectMeta
+
+    // Optional: Parameters is an array of Parameters used during the
+    // Template to Config transformation.
+    Parameters []Parameter
+
+    // Required: A list of resources to create
+    Objects []runtime.Object
+
+    // Optional: ObjectLabels is a set of labels that are applied to every
+    // object during the Template to Config transformation
+    // These labels are also be applied to selectors defined by objects in the template
+    ObjectLabels map[string]string
+}
+```
+
+**Parameter Object**
+```
+// Parameter defines a name/value variable that is to be processed during
+// the Template to Config transformation.
+type Parameter struct {
+    // Required: Parameter name must be set and it can be referenced in Template
+    // Items using $(PARAMETER_NAME)
+    Name string
+
+    // Required: Type-value of the parameter (one of string, int, bool, or base64)
+    Type ParameterType
```

Comment:
for case (1) we could be smarter in our substitution logic... we can easily check if, when we do the replacement, we're just setting the parameter value, or we're appending it to something else.  if we're appending, we can ensure the result is a string.  if we're not, we can ensure the result is whatever type we expected (which is really just a question of whether we include quotes or not).

does not cover case (2) (where a more complex conversion is required) or a case where someone wants to use an "int" parameter type as the entire value of a string field (they'd have to append a space or something to force it to be a string when we substitute).

but i'm not sure how probably those latter scenarios area...wondering if my solution to (1) is sufficient to make this acceptable (with the benefit of a simpler syntax).